### ISSUE TITLE
Raw links no longer working w/Dec14th update (0.1.5)

### ISSUE THREAD

**Issue by CollinChaffin** [Mottie/GitHub-userscripts#34]
Hey Rob!

I see the update on Dec 14th seems to indicate using a modifier, but all I see now when I click the raw links is a spinning Github logo and the links never appear.  Since it was just updated 2 weeks ago, am I just maybe missing something new I need to do?  I'm trying to go through the code now I really did not compare the changes with the last update since I always trust your scripts......

Thanks!

Collin

**Comment by Mottie** [comment]
Hi @CollinChaffin!

Hmm, I wonder if you've just hit the limit on API calls for the hour? It's working fine for me... I guess maybe I need to add a message if there is a problem.

![dev2](https://user-images.githubusercontent.com/136959/34456546-c5d21202-ed5e-11e7-93d6-9361f588ac5b.gif)

**Comment by CollinChaffin** [comment]
Hi did you see my edit?  It works fine when I disable Github Dark (StylishThemes) which is a pretty significant script.  It always worked without issue along side it (and several others from both you and StylishThemes) until this last update.  Looking at the code, nothing immediately jumps out at me, but I also haven't been living inside the Github API's etc. like you guys so maybe if you take a look it will jump out at you as something simple.  But again I have now re-tested with ONLY those 2 scripts enabled and it fails when the Github Dark is enabled, and immediately upon disabling that one and refreshing, yours begins working (but I lose all that other functionality).  Since that one is so popular, I'm guessing I won't be the only one running into this issue I just hope enough are also using the RawLinks that it would make sense to ask you to take a closer look at maybe making them work together (again).  (Of course, it is the holidays and you could be overcome with holiday spirit....) LOL :)

**Comment by Mottie** [comment]
Hehe, I wrote *most* of the GitHub dark userscript as well.

What browser are you using? I tested with both scripts running in Chrome and Firefox and I still can't reproduce the issues you're describing. Are you seeing any errors in the console? What about the network tab?

**Comment by CollinChaffin** [comment]
Sorry, meant to add the 2 captures I just did with the only change being disabling the Github Dark userscript in between and a page refresh:

With 2 scripts enabled - this Github Raw Links AND the Github Dark by StylishThemes:


![snag-12-30-201712 59 27-pm](https://user-images.githubusercontent.com/9301483/34456698-ca971950-ed62-11e7-999b-c5ece005d623.gif)

After disabling the Github Dark userscript and a page refresh:

![snag-12-30-20171 00 01-pm](https://user-images.githubusercontent.com/9301483/34456699-e1421e84-ed62-11e7-8407-e99a5766b3e8.gif)

**Comment by CollinChaffin** [comment]
LOL I swore you wrote that one that's why I was a bit baffled!  Well then, hopefully it really could come to you with a glance at the code(s)! :)

**Comment by CollinChaffin** [comment]
Okay so I think I found the issue.....

I added 2 console.log points at lines 115 and 117 to verify whether the inner loop that actually populates the dropdown is firing in both cases with and without Github Dark.

Then, a DOH! moment....for some reason, ONLY when Github Dark is enabled......here is the difference:

First, the successful run with Github Dark DISABLED in the console:
![snag_12-30-2017_13-31-25](https://user-images.githubusercontent.com/9301483/34456808-ef2cbd12-ed65-11e7-955c-de7fde4e9a5d.png)

Then, let's break it by enabling Github Dark again:
![snag_12-30-2017_13-31-00](https://user-images.githubusercontent.com/9301483/34456813-fb6b0dcc-ed65-11e7-8657-a8a190a71eb4.png)

So, now it's clear......the required api.github.com is simply being blocked because it is not in the allowed @connect (I'm assuming in Github Dark).......so I added this line INTO THE LATEST GITHUB DARK (it is already enabled in this script):

// @connect     api.github.com

....and the problem is resolved!

So, if you have control over that one as well, it should probably already have had api.github.com added to it's allowed @connect.....but now it looks like because of it's hooks it's pretty much required when run with these helpers.

I'll let you do any verification yourself and then close this issue when you're comfortable!

**Comment by Mottie** [comment]
Oh cool!

I just updated the raw-links script to show any error messages in the popup. I'll take a look at the GitHub-Dark script now.

Do you use [Stylus](https://github.com/openstyles/stylus)? And have you seen the [GitHub Dark usercss style](https://github.com/StylishThemes/GitHub-Dark#installing) (you can install it *directly* from our repo)?

**Comment by Mottie** [comment]
Hmm, actually that's odd... the GitHub Dark userscript doesn't use the `api.github.com`, but the raw gist userscript does.

Maybe it's an issue with the userscript manager you're using. Which one are you using and what browser?

**Comment by CollinChaffin** [comment]
Yes, thanks but I need all the functionality of GM/TM above and beyond the CSS of Stylus but I love the fact that it's returned to open source and away from the Stylish!  I'm guessing because of the hooks that the Dark script is making and loading the same prereqs that it is causing the issue but adding that allow to it and it's working great together again regardless of load order etc.

**Comment by CollinChaffin** [comment]
Tampermonkey latest on Chrome x64 (I added it above in an edit for you).  If I have time I'll debug deeper to figure out exactly why and I agree, clearly Github Dark isn't using it or it would be broken, but it does so much hooking itself all I can think is if you think about it all the calls on that gist page that Dark makes and overrides default, now Rawlinks comes along and also wants to override a default that I'm guessing has already been touched by Dark.......and maybe the Rawlinks load call forces the Dark script override to fire now calling itself on API which then hits a wall since it's now allowed and looks like IT is the caller.......just a high level theory.

**Comment by darkred** [comment]
I use the usercss version with Stylus (not the userscipt).

Nevertheless, after I read the OP, I  checked and found that I had the same issue: a spinning Github logo and the links never appear. 
So, by toggling a few of the 22 scripts that run in gist pages -[example](https://gist.github.com/BrockA)- (most of which written by Mottie) 
I quickly found that, the script that would conflict with the 'Gist Raw Links' script 
was the [translate.google tooltip](https://greasyfork.org/en/scripts/9285-translate-google-tooltip), 
and so, by adding `https://gist.github.com\*` to the ''Gist Raw Links' script's Settings | User Excludes 
the issue is fixed for me. 
(Using Chrome 63 and Tampermonkey 4.4)

**Comment by Mottie** [comment]
Thanks for the feedback. It's odd that I can't reproduce the problem and I too am using Chrome 63, but I'm using TM beta v4.5.

**Comment by CollinChaffin** [comment]
LOL thanks @darkred  for the info but I'm confused if you exclude gist.github.com, you are excluding the actual source site required by the GIST raw links. :)

Also, if you look above at my console screengrab you can see at least in my case, the debug clearly shows that the CALLING script was Github Dark (when enabled) being forced by the Rawlinks when the call is made since GH Dark has overridden the default handlers (when enabled).  At no point (at least for me) does Google Translate interfere I have it enabled and have tried with it enabled and disabled.....so it sounds like perhaps a secondary CSS-only issue you're facing since Tthe javascript console clearly shows the security disallow blocking the required call until manually added.  Once I add the @connect line to GH Dark, even with Google Translate and over 200 other TM scripts re-enabled, it again works 100% of the time without issue.

@Mottie to clarify, I am running the following:

Chrome:                             65.0.3294.5 (Official Build) dev (64-bit)
Tampermonkey:                4.4
OS:                                    Windows 7 x64

Let me see if I can add some additional breakpoints to get a better handle on where it actually pushing it back to GH Dark to fail......

**Comment by darkred** [comment]
> LOL thanks @darkred for the info but I'm confused if you exclude gist.github.com, you are excluding the actual source site required by the GIST raw links. :)

Typo.

**Comment by Mottie** [comment]
The screenshot is showing the error from script `VM590:36` - a different script than the `VM604` showing the "HIT" messages. Either way, if two different userscripts are causing the same issue, I think it might be related to TM and not either script.

Maybe @derjanb can verify if this is an problem?

**Comment by CollinChaffin** [comment]
What further proves the async hooks from GH Dark are making the GM.xmlHttpRequest, you can actually leave everything broken (in my tests) and simply change this Rawlinks script from @run-at document-IDLE to document-START (which is what GH Dark is), and bang it also fixes the issue.  That has slipped by me originally which is also why changing the load order in TM was irrelevant, since the GH Dark at doc-start will always win, anyway.

When I log the call from line 89 which ultimately makes the http external call, it fires and debug logs either way, but when set to the current default of doc-idle, it logs the call and hands off to the async GH-Dark which apparently then makes the GM.xmlHttpRequest because it's sitting in an async loop for it's events then logs the security allow error.  Not sure why on earth that behavior would change between 4.4 and 4.5 beta as you indicated, unless something with async GM.xmlHttpRequest has inherently changed?  There are so many async event listeners overriding the defaults in GH Dark but regardless it's a big concerning to me either way that at doc-idle GH Dark is still doing ANYTHING regarding GM.xmlHttpRequests......it should be loading and getting out of the way unless an event fires, right?

I tried uncommenting the end debug block in GH Dark but it literally throws so many hard errors in the console I'm happy to try to help but I would need some guidance as to how to properly even debug that script using any built-in debug routines.

**Comment by CollinChaffin** [comment]
@Mottie VM59 is GH Dark - I only have 2 enabled.  VM604 is this script and that's my point - GH Dark is sitting in async loop and it is VM590 which is returning the security error and also why that line VM590 disappears and the calls go through when I add the @connect line to the GH DARK script.  Sorry I must not have explained that clearly before!

**Comment by Mottie** [comment]
No, you were clear. I just don't see how adding a `// @connect api.github.com` to the GitHub Dark Script since that script is only accessing `raw.githubusercontent.com` and does have a corresponding `// @connect raw.githubusercontent.com`.

**Comment by CollinChaffin** [comment]
It's not.  Ultimately there is/was already an underlying issue with GH Dark that I never noticed that is causing it to never init.  On multiple machines, if I sit and refresh the page and click the dropdown where GH Dark appends the "Github Dark Settings" for example, 50% of the time it never appears due to the code not completing.  I'm convinced the same issue is basically making the async (don't need to tell you since you wrote it) that encapsulates the entire thing and then all the async calls inside for the ajax.....simply loop never get to the full init and hence since that script has registered via this:

on($(".ghd-update", panel), "click", async event => {
      event.preventDefault();
      await forceUpdate();
    });

...etc. and grabbed hold of everything, IT is what then places the xmlhttprequest......again I agree don't know exactly why I've got 2 little ones running around and trying to debug as I have time, but the reason this issue 100% goes away completely just by forcing RawLinks to run at doc-startup and not idle is that then IT makes the xmlhttprequest as it should, and immediately populates the list......and then Dark loads after it (since I've got Rawlinks first).

But remember.....if I can literally disable all but ONE script - the GH Dark, and click refresh on Github and half the time it not properly register/add itself to the user dropdown menu at all render it's CSS colors etc. but NOT properly register the element changes.....no need to even do much more testing with the Rawlinks script which I think ultimately is 100% issue-free after all of this testing.....just my opinion but I'll keep trying to help get to the bottom of it since GH Dark is actually my #1 favorite/most used script!

**Comment by Mottie** [comment]
Ok, I just updated the GitHub Dark script to not rely on the DOMContentLoaded event. Please let me know if that fixes the issue.

**Comment by derjanb** [comment]
@Mottie @CollinChaffin I've not checked it in detailt, but I think the issue is caused by [requiring the GM4 polyfill](https://github.com/Mottie/GitHub-userscripts/blob/master/gist-raw-links.user.js#L15) which leaks the GM object to the global scope. 
If now there are two scripts using it, then only one script will be able to use the right one with the right `@connect` restrictions. Can you please check if the fix from here solves the issue? https://github.com/greasemonkey/greasemonkey/issues/2784#issue-285186843

Thanks.

**Comment by Mottie** [comment]
I still can't duplicate the issue 😿
